### ISSUE TITLE
toMany in POST and PUT throwing unrecognizedIdentifierFields Exception

### ISSUE THREAD

**Issue by nobesnickr** [zfcampus/zf-apigility-doctrine#183]
Greetings folks,

I am seeing a weird issue that I'm not certain if is either a bug in this module, the DoctrineObject hydrator or something I am doing wrong / configured incorrectly. 

What I am trying to do is either a POST or PUT (both have same effect) with a collection for one of the fields. The goal is to handle adding/removing/updating the entities in the collection along with the parent entity (Adding/Removing handled by Hydration Strategies). 

My specific use case is managing Phone Numbers for a Person.

So for example, my form POST's a JSON object such as:
```
{
    "firstName": "John",
    "lastName": "Doe",
    "phoneNumbers": [
        {
            "phoneNumber": "+14805551234"
        },
        {
            "phoneNumber": "+14805557777"
        }
    ]
}
```

or PUT's a JSON object such as:
```
{
    "id": 123,
    "firstName": "John",
    "lastName": "Doe",
    "phoneNumbers": [
        {
            "id": 1,
            "phoneNumber": "+14805551234"
        },
        {
            "id": 2,
            "phoneNumber": "+14805557777"
        }
    ]
}
```

My input filter is setup to handle and validate the collection already but when the entity goes to hydrate the toMany association I get an error of: 
```
{
    "file": "/vagrant/vendor/doctrine/orm/lib/Doctrine/ORM/EntityManager.php",
    "line": 411,
    "function": "unrecognizedIdentifierFields",
    "class": "Doctrine\\ORM\\ORMException",
    "type": "::",
    "args": [
        "MyAwesomeApp\\Entity\\PhoneNumber",
        [
            "phoneNumber"
        ]
    ]
}
```

I've already configured my entity to handle the collection on the add / remove methods for the collection but its never being called. The odd thing is, the concept works flawlessly when using Zend\Form and a Collection Fieldset (I'm assuming because the fieldset returns an entity instead of an array).

If this is a bug, or enhancement or anything that requires work I'd be more than happy to handle it and push it up. If its not and anyone could point me in the right direction it would be most appreciated!

**Comment by TomHAnderson** [comment]
Well, my first thought is you shouldn't be calling colletions plural.  The field should be phoneNumber[] but maybe I'm splitting hairs.  As long as you're sure you're setting the field correctly I think what you need is a hydrator strategy.  https://github.com/TomHAnderson/zf-apigility-doctrine/blob/master/src/Server/Hydrator/Strategy/CollectionExtract.php says hydration is not supported for collections but check this out: https://github.com/TomHAnderson/RollnApi/blob/master/module/RollNApi/src/RollNApi/Hydrator/Strategy/Performer/Album.php config: https://github.com/TomHAnderson/RollnApi/blob/master/module/RollNApi/config/module.config.php#L419

So where the many to many contains the id of the other entity you may join them, create them, maybe, etc.

**Comment by nobesnickr** [comment]
Tom, Thank you so much for the quick and thorough feedback. I think I might not have properly specified my particular issue or it may just be that I'm going about it the wrong way (depending on a bit too much "magic" for a restful endpoint). 

First off, yes the hydrating strategy works great for adding/removing entity relationships, however isn’t exactly what I’m trying to accomplish. What I am trying to avoid is implementing late arriving dimensions on the related entities during the creation of a new "parent" entity, and for updates of an existing parent entity, cascade changes to the sub/related entities. The hydrator strategies trigger after the DoctrineObject association building (which makes sense but blocks the intercept on my side).

Allow me to explain my situation a little further and the two potential solutions to handle it:

A form to create a new "Person" entity can be simplified as a form with 2 fields and a Collection Fieldset that can have many "Phone Numbers". In this specific example we’ll assume the 2 Phone Numbers are new in the system. The resulting JSON of the form would give me output referenced above. 
To handle this in a pure RESTful context I could implement 3-5 different calls which would:
	1.	Fetch PhoneNumber[0]  to see if it exists in the system, if so store the id in place of the object Person.PhoneNumbers[0]
	2.	If PhoneNumber[0] doesn’t exist, POST and store the new id in place of the object at Person.PhoneNumbers[0]
	3.	Repeat Step 1 for PhoneNumber[1]
	4.	Repeat Step 2 for PhoneNumber[1]
	5.	POST the Person object

The other solution I am hoping/looking for is being able to handle these relationships in one request. I’ve gotten the content-validation setup to handle and validate the collection so I’m confident by the time the hydrator is firing we’re only dealing with valid data. 

Perhaps I’ve been spoiled by the current usage of Form and Fieldsets which handles this concept in one request or asking to much of the "stock" DoctrineObject hydrator.

**Comment by TomHAnderson** [comment]
Ok, allow me to move the conversation in a different direction.  
The other day I tried to create a listener for doctrine events on the zf-apigility-doctrine DoctrineResource.php.  Here's what I created: https://github.com/TomHAnderson/apigility-oauth2-doctrine-skeleton/blob/master/module/Api/src/Api/EventListener/OAuth2/Client/CreateListener.php

Because of my work there I created a new pull request and changed the order of event triggers in the DoctrineResource.  This is still in PR waiting for a review and I think we can solve your problem by adding to the doctrine event triggers.  https://github.com/TomHAnderson/zf-apigility-doctrine/blob/hotfix/doctrine-events/src/Server/Resource/DoctrineResource.php

Where the call is generally 
```
        $results = $this->triggerDoctrineEvent(DoctrineResourceEvent::EVENT_FETCH_POST, $entity);
```

I think we should change the third parameter from collection (not used after my changes) to the original data sent to the DoctrineResource $data.  If you have that data would being able to hook into the events resolve what you're trying to do?  We'll extend the PR to cover your needs right now and now is the right time to add to the PR.  These settings haven't changed in a year :+1:

**Comment by veewee** [comment]
@TomHAnderson 
Adding the $data is surely a good thing to do!
For update/patch events you might want to create a DoctrineResourceChangedEvent which has the extra method: getOriginalData(). This will load the original entity data from the Doctrine UnitOfWork.
This can be handy to track and validate the changes in an entity.
For example: if status was set to 'ok', it cannot be set to 'not ok', but a status can be set to 'not ok' in any other situation.

**Comment by nobesnickr** [comment]
Wonderful! I'll check that out and update with my findings.

Is this concept something that would be helpful to the community? If so I'd be more than happy to take the helm on flushing out a more generic implementation and getting it PR'd.

**Comment by TomHAnderson** [comment]
@nobesnickr Yes, this is something the community needs.  When it was done the first time the workflow wasn't defined at all so it was implemented in a little hap-hazard way.  I tried to fix that with #182  Push a [wip] PR to zf-apigility-doctrine so we can watch your work from there please.  If you can start your PR with merging #182 I'll watch your work eagerly.

**Comment by nobesnickr** [comment]
Sounds good, I just pulled your branch so let me kick it around this weekend and I'll push up a PR when I have something tangible.

**Comment by nobesnickr** [comment]
@TomHAnderson Before I start ripping things apart I want to make sure I'm at the right starting point. Were you referencing that you already have a branch replacing Collection with original data or I should start there (I don't want to squash/rework if you do and didnt see it in hotfix/doctrine-events)?  

Also, I may need to add pre/post hydrate events which is where I believe my blocker originates from, any issue with that?

**Comment by nobesnickr** [comment]
I just pushed an update to the main event handler (CollectionEvent) which allows for recursively nested collections and made it a bit more testable. 

The one issue however is that the ZF2 Input Filter Factory doesn't currently support nested CollectionInputFilter so I had to patch mine locally, I'll open a PR on the main ZF2 repo for the few lines I had to change. This shouldn't affect this code at all, as the IV is upstream so the recursive functionality will simply be unavailable until that change gets implemented in the ZF2 repo.
